# Исследование надёжности заёмщиков банка

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание проекта</a></span></li><li><span><a href="#Открытие-и-изучение-данных" data-toc-modified-id="Открытие-и-изучение-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Открытие и изучение данных</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Заполнение-пропусков" data-toc-modified-id="Заполнение-пропусков-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Заполнение пропусков</a></span></li><li><span><a href="#Проверка-данных-на-аномалии-и-исправления" data-toc-modified-id="Проверка-данных-на-аномалии-и-исправления-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Проверка данных на аномалии и исправления</a></span></li><li><span><a href="#Изменение-типов-данных" data-toc-modified-id="Изменение-типов-данных-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Изменение типов данных</a></span></li><li><span><a href="#Удаление-дубликатов" data-toc-modified-id="Удаление-дубликатов-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Удаление дубликатов</a></span></li><li><span><a href="#Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма" data-toc-modified-id="Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма</a></span></li><li><span><a href="#Категоризация-дохода" data-toc-modified-id="Категоризация-дохода-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Категоризация дохода</a></span></li><li><span><a href="#Категоризация-целей-кредита" data-toc-modified-id="Категоризация-целей-кредита-3.7"><span class="toc-item-num">3.7&nbsp;&nbsp;</span>Категоризация целей кредита</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.8"><span class="toc-item-num">3.8&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Ответы-на-вопросы-заказчика" data-toc-modified-id="Ответы-на-вопросы-заказчика-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Ответы на вопросы заказчика</a></span><ul class="toc-item"><li><span><a href="#Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Есть ли зависимость между количеством детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-4.6"><span class="toc-item-num">4.6&nbsp;&nbsp;</span>Вывод:</a></span></li><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-4.7"><span class="toc-item-num">4.7&nbsp;&nbsp;</span>Как разные цели кредита влияют на его возврат в срок?</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.8"><span class="toc-item-num">4.8&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

## Описание проекта

В данном проекте изучается надёжность клиентов-заёмщиков банка.   

**Заказчик исследования** - кредитный отдел банка.

**Основные задачи**:

 - Подготовить данные для построения модели кредитного скоринга - системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

- Выяснить, какие факторы и как влияют на потенциальную платёжеспособность клиентов, какие категории клиентов наиболее вероятно погасят кредит в срок, а какие категории клиентов вероятнее могут иметь просрочки по платежам. Заказчику важны ответы на следующие вопросы:
     - есть ли зависимость между количеством детей и возвратом кредита в срок?
     - есть ли зависимость между семейным положением и возвратом кредита в срок?
     - есть ли зависимость между уровнем дохода и возвратом кредита в срок?
     - как разные цели кредита влияют на его возврат в срок?

**План работы**:
- открытие и знакомство с данными
- предобработка данных
- подготовка данных для передачи их специалисту data science для построения модели кредитного скоринга
- анализ данных - ответы на вопросы заказчика
- общий вывод


**Описание данных**:

Данные предоставлены банком, это статистика платёжеспособности клиентов.



## Открытие и изучение данных

In [1]:
import pandas as pd

In [2]:
data = pd.read_csv('/datasets/bank_data.csv')

Посмотрим, как выглядит таблица с данными и общая информация о датасете.

In [3]:
display(data.head()) 

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

In [4]:
data.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


На всякий случай продублируем "сырые" данные, чтобы в дальнейшем, после предобработки, можно было сравнить два датасета.

In [5]:
data_raw = pd.read_csv('/datasets/bank_data.csv')

### Вывод

В данных 21525 строк, 12 столбцов.

На первый взгляд, здесь есть следующие недочёты:

* судя по информации о данных, **есть пропуски** в разделах: 
    - 'days_employed' - общий трудовой стаж в днях,
    - 'total_income' - ежемесячный доход.
* есть **аномальные значения** в разделе 
    - 'days_employed' - общий трудовой стаж в днях записан отрицательным числом
* есть данные, которые нужно **привести к одному регистру** - в графе
    - 'education' - образование

## Предобработка данных

Предобработка включает в себя следующие разделы:
- заполнение пропусков
- проверка данных на аномалии и исправления
- изменение типов данных
- удаление дубликатов
- формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма
- категоризация дохода
- категоризация целей кредита
- вывод


### Заполнение пропусков

In [6]:
display(data[data['total_income'].isna()].head()) 

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

    total_income                           purpose  
12           NaN                   сыграть свадьбу  
26           NaN                       образование  
29           NaN  строительство жилой недвижимости  
41           NaN  сделка с подержанным автомобилем  
55           NaN                   сыграть свадьбу

Пропуски представлены в виде **NaN**.

In [7]:
percentage_of_missing_values = data['total_income'].isna().mean()
print(f'Доля пропущенных значений составляет: {percentage_of_missing_values:.0%}')

Доля пропущенных значений составляет: 10%


В данных об **общем трудовом стаже в днях** есть пропуски. Возможно, это связано со сложностью уточнения таких данных самим клиентом:
- менялось законодательство, в разные годы трудовой стаж считался по-разному, в частности для людей старшего поколения декретный отпуск мог входить или не входить в общий стаж, 
- не все документы, подтверждающие стаж, могут быть оформлены правильно или быть в наличии у клиента,
- клиент мог часто менять работу, чередовать официально оформленную занятость в организации и работу частным образом (не оформленную и не задокументированную)

Также пропуски есть и в данных об **ежемесячном доходе**. Это может быть связано со следующими факторами:
- нежелание клиента афишировать свой доход (из опасений о личной безопасности и безопасности таких данных, если доход высокий),
- сложностями расчёта такого дохода, если клиент менял работу/какое-то время был безработным/попал в сложную жизненную ситуацию,
- имеются сложности с расчётом среднего дохода и официальным подтверждением его (например, клиент предпочитает работу частным образом, доход сильно меняется от месяца к месяцу, нет в наличии документов, которые могли бы подтвердить доход за последние месяцы)
- у клиента есть официально не оформленные источники дохода и поэтому он стремится не указывать реальное положение своих финансовых дел

**Вывод**: 
- возможно, мы имеем дело с сознательным пропуском клиентом заполнения этих разделов данных о себе
- возможно имеет место технический сбой (настораживает абсолютно идентичное количество пропусков в обоих столбцах)

In [8]:
data['total_income'] = data['total_income'].fillna(data.groupby(['income_type'])['total_income'].transform('median'))

In [9]:
data['total_income'].isna().sum()

0

Теперь все **пропуски** в столбце **с данными о ежемесячном доходе заполнены** медианным значением **в зависимости от типа занятости клиента** (и можно убедиться в этом, снова посчитав пропуски в этом столбце - их 0).

**Медианное значение** в этом случае даёт более объективную информацию (по сравнению со средним арифметическим). Это особенно актуально в странах, где есть большой разрыв в доходах населения, и могут быть выдающиеся от средней нормы значения, которые и уведут общую картину в необъективную сторону. Поэтому с этой количественной переменной при заполнении пропусков лучше использовать именно метод median().

Также есть пропуски в столбце с данными о трудовом стаже, но этот столбец имеет смысл рассмотреть отдельно в следующем шаге, т.к. в нём замечены аномалии, и прежде чем заполнять эти пропуски, надо выяснить природу аномалий и общее содержимое столбца.

### Проверка данных на аномалии и исправления

In [10]:
display(data.sample(5, random_state=42))

children  days_employed  dob_years education  education_id  \
423           0    -191.167760         56   среднее             1   
3522          0   -2319.817259         58   среднее             1   
8760          1   -2990.578297         34    высшее             0   
20695         0    -926.209452         28   среднее             1   
4351          0   -2524.302106         42   СРЕДНЕЕ             1   

               family_status  family_status_id gender income_type  debt  \
423          женат / замужем                 0      M   сотрудник     0   
3522   Не женат / не замужем                 4      F   сотрудник     0   
8760        гражданский брак                 1      M   компаньон     0   
20695  Не женат / не замужем                 4      F   сотрудник     0   
4351         женат / замужем                 0      F   сотрудник     0   

        total_income                        purpose  
423    138653.748793                     автомобили  
3522   123152.627177                  покупка жилья  
8760   232380.737167                        свадьба  
20695  210617.086601  получение высшего образования  
4351    78487.540219           сделка с автомобилем

Ещё раз взглянув на таблицу, займёмся следующей явной проблемой: в столбце 'days_employed' есть **аномальные значения: общий трудовой стаж в днях часто записан в виде отрицательного числа**. 

In [11]:
days_employed_minus = data['days_employed'] < 0
print(f'Количество аномальных отрицательных значений: {days_employed_minus.sum()}')

Количество аномальных отрицательных значений: 15906


In [12]:
days_employed_plus = data['days_employed'] > 0
display(data.loc[days_employed_plus].sort_values(by='days_employed').tail()) #смотрим на конец отсортированной таблицы

children  days_employed  dob_years education  education_id  \
7794          0  401663.850046         61   среднее             1   
2156          0  401674.466633         60   среднее             1   
7664          1  401675.093434         61   среднее             1   
10006         0  401715.811749         69    высшее             0   
6954          0  401755.400475         56   среднее             1   

               family_status  family_status_id gender income_type  debt  \
7794        гражданский брак                 1      F   пенсионер     0   
2156         женат / замужем                 0      M   пенсионер     0   
7664         женат / замужем                 0      F   пенсионер     0   
10006  Не женат / не замужем                 4      F   пенсионер     0   
6954          вдовец / вдова                 2      F   пенсионер     0   

        total_income                purpose  
7794    48286.441362                свадьба  
2156   325395.724541             автомобили  
7664   126214.519212      операции с жильем  
10006   57390.256908  получение образования  
6954   176278.441171           ремонт жилью

Кроме отрицательных значений, есть и другие аномалии - неправдоподобно высокие показатели. Например, чтобы в стаже было 
401755 дней, нужно было бы работать без выходных 1128 лет подряд.

Посмотрим, сколько таких аномально высоких значений. Для расчётов возьмём среднее количество рабочих дней в году (247)
и максимально возможное реалистичное количество лет стажа (80).


In [13]:
days_employed_plus_anomal = data['days_employed'] > 247*80
print(f'Количество аномальных положительных значений: {days_employed_plus_anomal.sum()}')

Количество аномальных положительных значений: 3445


In [14]:
print(f'Общее количество аномальных значений: {days_employed_minus.sum() + days_employed_plus_anomal.sum()}')

Общее количество аномальных значений: 19351


Получается, **все данные** в колонке про общий трудовой стаж в днях - **аномальны и не пригодны для анализа**:

- бОльшую часть составляют **аномальные отрицательные значения**,
- также есть и **неправдоподобно высокие значения**,
- и, кроме того, есть **пропуски** (как и в столбце про ежемесячный доход, 10% от общего числа значений).

Судя по всему, есть **проблема в сборе данных**. В чём она может быть с точки зрения клиента:

- при заполнении со стороны банка указан не очень удобный или **неграмотный шаблон** для подсчёта стажа по дням, **или** возможно, **шаблона** и вовсе **нет**, и пользователям не на что опереться, отсюда возникает массовая путаница: 
   * клиенты могут делать расчёт правильно, но в конце по невнимательности указывают минус перед числом, видя в шаблоне дефис (в результате получается отрицательное число),
   * клиенты вычитают из даты начала трудовой деятельности дату конца и в результате в расчёте получается отрицательное число,
   
   * клиенты делают расчёт стажа в часах, а не в днях, и отсюда возникают нереалистично высокие цифры,
   * клиенты неправильно переводят время стажа из лет и месяцев в дни (умножают и на количество дней в году, и на количество рабочих дней в году)
   * клиенты на каком-то шаге вместо складывания различных отрезков стажа между собой умножают их.


Для целей нашего исследования данные про стаж не являются принципиально важными, 
поэтому на данном этапе мы просто заполним пропуски.
Если появится задача/гипотеза, напрямую связанная с трудовым стажем, то следует **обратиться к разработчикам** и наладить **сбор корректных данных** от клиентов.

In [15]:
days_employed_median = data['days_employed'].median()
data['days_employed'] = data['days_employed'].fillna(days_employed_median)

data['days_employed'].isna().sum()

0

Пропуски заполнены медианным значением.
Код работает - теперь сумма пропусков в этом столбце равна 0.

Теперь посмотрим, есть ли ещё аномалии в данных. Пройдёмся по всем столбцам по порядку.

In [16]:
data['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Очевидно, и сюда закрались **аномалии**:
- **20 детей** - это слишком большой отрыв от реальности, когда мы видим, что количество детей очень плавно снижается при движении от 0 к 5 (а значений 6, 7 и т.д. попросту нет). Скорее всего, значение 2.0 потеряло точку (технический сбой) или дело в опечатке; в любом случае логичнее всего предположить, что имеется ввиду значение 2.
- **-1 ребёнок** - скорее всего, это тоже опечатка от значения 1, просто в данные попал дефис (либо ребёнок погиб, и это записали таким образом; или так обозначено отсутствие детей. В качестве рабочей выбираю первую версию).

In [17]:
data['children'] = data['children'].replace(20, 2) #заменяем опечатки
data['children'] = data['children'].replace(-1, 1) #заменяем опечатки
data['children'].value_counts() #проверяем, всё ли теперь в порядке

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

Теперь данные столбце 'children' будут пригодны для анализа. Следующий столбец -'days_employed' - уже анализировали.
Переходим к 'dob years' - возраст клиента в годах.

In [18]:
data['dob_years'].value_counts()

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
22    183
66    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

Здесь виден один артефакт - 101 клиент с возрастом 0. Посмотрим на такие строки поближе.

In [19]:
data.loc[data['dob_years'] == 0].sample(5, random_state=42)

children  days_employed  dob_years education  education_id  \
16861         0  395893.813367          0   среднее             1   
11481         2   -1611.360609          0   среднее             1   
13232         0    -229.560933          0    высшее             0   
13413         0    -947.275258          0    ВЫСШЕЕ             0   
10163         0   -4847.111204          0   Среднее             1   

          family_status  family_status_id gender  income_type  debt  \
16861         в разводе                 3      F    пенсионер     0   
11481   женат / замужем                 0      M    сотрудник     0   
13232   женат / замужем                 0      M    сотрудник     0   
13413   женат / замужем                 0      M    компаньон     0   
10163  гражданский брак                 1      F  госслужащий     0   

        total_income                           purpose  
16861  218284.757642                     покупка жилья  
11481  236698.919494  строительство жилой недвижимости  
13232  131782.794583             заняться образованием  
13413  182747.326874      на покупку своего автомобиля  
10163   97169.690307                           свадьба

Судя по прочим данным, 0 - это не реальный возраст клиентов (младенец не может иметь высшего образования или быть в разводе), 
а по сути пропуск таких данных. Заменим этот артефакт на среднее арифметическое по этому столбцу (здесь нет выдающихся значений, метод mean() должен показать объективный результат).

In [20]:
dob_years_mean = data[data['dob_years'] > 0]['dob_years'].mean() #считаем среднее арифм. по столбцу без учёта артефактов

In [21]:
data.loc[data['dob_years'] == 0, 'dob_years'] = dob_years_mean #заменяем 0 на среднее арифметическое

Проверим на наличие аномалий оставшиеся столбцы.

In [22]:
data['education'].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

Здесь нужно будет привести запись значений к одному регистру (работа для следующего раздела), аномалий по количеству значений не наблюдается (например, было бы странно, если все были с учёными степенями или наоборот, только с начальным образованием).

In [23]:
data['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

Тоже всё правдоподобно, проблем не видно.

In [24]:
data['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

In [25]:
display(data[data['gender'] == 'XNA'])

children  days_employed  dob_years            education  education_id  \
10701         0   -2358.600502       24.0  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10701  гражданский брак                 1    XNA   компаньон     0   

        total_income               purpose  
10701  203905.157261  покупка недвижимости

В столбце 'gender' есть один артефакт. Происхождение его неизвестно. 
Какие версии мне пришли на ум:
- опечатка, клиент писал на русской раскладке алфавита (муж или жен) - не соответствует действительности (XNA - это ЧТФ),
- это наименование нового гендера (поискала в интернете, не нашла таких вариантов). В поиске это скорее связано с Microsoft XNA (Набор инструментов с управляемой средой времени выполнения, созданный Microsoft, облегчающий разработку и управление компьютерными играми) 

Обе версии, очевидно, не подходят. Аномалия может быть вызвана как человеческим фактором, так и техническим сбоем.
На цели исследования этот артефакт никак не повлияет (поэтому просто заменим его на более понятный 'unknown').

In [26]:
data.loc[data['gender'] == 'XNA', 'gender'] = 'unknown'

In [27]:
data['gender'].value_counts()

F          14236
M           7288
unknown        1
Name: gender, dtype: int64

In [28]:
data['income_type'].value_counts()

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
студент                1
в декрете              1
Name: income_type, dtype: int64

In [29]:
data['debt'].value_counts()

0    19784
1     1741
Name: debt, dtype: int64

In [30]:
data['purpose'].value_counts()

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
покупка жилья для сдачи                   653
операции с жильем                         653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
строительство недвижимости                620
покупка своего жилья                      620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

В оставшихся столбцах аномалий больше не найдено, но в столбце 'puprose'(цель получения кредита) есть много неявных дубликатов.

### Изменение типов данных

В столбце 'total_income' данные представлены дробным числом (тип данных float64), заменим его для удобства чтения таких сведений на обычные целые числа.

In [31]:
data['total_income'] = data['total_income'].astype('int')

### Удаление дубликатов

Итак, подошла пора заняться дубликатами. Они имеются в двух столбцах: 'education' и 'purpose'.
Вспомним, как выглядят дубликаты в графе об образовании.

In [32]:
data['education'].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

Приводим все строки к нижнему регистру и сохраняем такой формат записи в таблице. Затем проверяем, всё ли в порядке.

In [33]:
data['education'] = data['education'].str.lower() 
data['education'].value_counts() 

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

Проверяем, есть ли ещё дубликаты, не в отдельном столбце, а во всей таблице, и сколько их.

In [34]:
data.duplicated().sum() 

71

Убираем строки-дубликаты с заменой индексов, затем проверяем, всё ли сработало.

In [35]:
data = data.drop_duplicates().reset_index(drop=True) 

In [36]:
data.duplicated().sum() 

0

Я избавилась от дубликатов в столбце с данными об образовании с помощью приведения всех строк к одному регистру (метод .str.lower()), так как природа возникновения дубликатов была очевидна после вывода на экран value_counts(). 

Скорее всего, у клиентов не было единого шаблона, наглядного образца, как следует заполнять эту графу, и поэтому кто-то писал с Заглавной буквы, кто-то - с маленькой, а кто-то ВОТ ТАК. Если бы до или во время заполнения графы подсвечивался бы образец написания (как бывает на некоторых сайтах, например, формат записи телефона или электр.почты), то, возможно, такой проблемы удалось бы избежать. С этой идеей можно обратиться к разработчикам.

Также в таблице оказалась 71 строка с явными дубликатами (это дубликаты показал метод duplicated(), их сумму посчитал метод sum()), я удалила лишнее с помощью метода drop_duplicates() и произвела замену индексов с помощью reset_index(drop=True).

Также есть неявные дубликаты в столбце 'purpose', но логика проекта предполагает работу с ними на одгом из последующих шагов.

### Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма

Создаём словарь для данных об образовании, удаляем в нём дубликаты и проверяем, всё ли в порядке.

In [37]:
education_dict = data[['education_id', 'education']] 
education_dict = education_dict.drop_duplicates().reset_index(drop=True)
education_dict 

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

Делаем то же самое для данных о семейном положении.

In [38]:
family_status_dict = data[['family_status_id', 'family_status']] 
family_status_dict = family_status_dict.drop_duplicates().reset_index(drop=True)
family_status_dict

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  Не женат / не замужем

In [39]:
data = data[['children','days_employed','dob_years','education_id','family_status_id','gender',
             'income_type','debt','total_income','purpose']]

In [40]:
data.head() 

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1   -8437.673028       42.0             0                 0      F   
1         1   -4024.803754       36.0             1                 0      F   
2         0   -5623.422610       33.0             1                 0      M   
3         3   -4124.747207       32.0             1                 0      M   
4         0  340266.072047       53.0             1                 1      F   

  income_type  debt  total_income                     purpose  
0   сотрудник     0        253875               покупка жилья  
1   сотрудник     0        112080     приобретение автомобиля  
2   сотрудник     0        145885               покупка жилья  
3   сотрудник     0        267628  дополнительное образование  
4   пенсионер     0        158616             сыграть свадьбу

Теперь таблица стала компактнее. Этот датафрейм будет удобно использовать в дальнейшем.

### Категоризация дохода

Напишем функцию, которая создаст категоризацию дохода

In [41]:
def total_income_category (total_income):
    """
    Возвращает категорию дохода по значению ежемесячного дохода total_income,
    используя следующие правила:
    - 'E', если доход  <= 30 тыс.руб.;
    - 'D', если доход от 30001 до 50 тыс. руб.;
    - 'C', если доход от 50001 до 200 тыс.руб.;
    - 'B', если доход от 200001 до 1 млн.руб.;
    - 'A', если доход более 1 млн.руб.
    """
    if total_income <= 30000:
        return 'E'
    if total_income <= 50000:
        return 'D'
    if total_income <= 200000:
        return 'C'
    if total_income <= 1000000:
        return 'B'
    return 'A'

Проверяем работу функции для разных чисел:

In [42]:
total_income_category(25000) 

'E'

In [43]:
total_income_category(235000) 

'B'

Функция работает корректно.

Добавляем столбец с категориями доходов и проверяем, как выглядит результат.

In [44]:
data['total_income_category'] = data['total_income'].apply(total_income_category) 
display(data.head()) 

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1   -8437.673028       42.0             0                 0      F   
1         1   -4024.803754       36.0             1                 0      F   
2         0   -5623.422610       33.0             1                 0      M   
3         3   -4124.747207       32.0             1                 0      M   
4         0  340266.072047       53.0             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C

In [45]:
data = data[['children','days_employed','dob_years','education_id','family_status_id','gender',
             'income_type','debt','total_income','total_income_category','purpose']]

Переставляем новый столбец поближе к "родственному" столбцу с данными о доходах. Для него можно тоже сделать словарь.

In [46]:
display(data.head()) 

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1   -8437.673028       42.0             0                 0      F   
1         1   -4024.803754       36.0             1                 0      F   
2         0   -5623.422610       33.0             1                 0      M   
3         3   -4124.747207       32.0             1                 0      M   
4         0  340266.072047       53.0             1                 1      F   

  income_type  debt  total_income total_income_category  \
0   сотрудник     0        253875                     B   
1   сотрудник     0        112080                     C   
2   сотрудник     0        145885                     C   
3   сотрудник     0        267628                     B   
4   пенсионер     0        158616                     C   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

Создаём словарь для данных о доходе, удаляем в нём дубликаты, проверяем результат работы.

In [47]:
total_income_dict = data[['total_income_category', 'total_income']]
total_income_dict = total_income_dict.drop_duplicates().reset_index(drop=True) 

In [48]:
data = data[['children','days_employed','dob_years','education_id','family_status_id','gender',
             'income_type','debt','total_income_category','purpose']]

In [49]:
display(data.head())

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1   -8437.673028       42.0             0                 0      F   
1         1   -4024.803754       36.0             1                 0      F   
2         0   -5623.422610       33.0             1                 0      M   
3         3   -4124.747207       32.0             1                 0      M   
4         0  340266.072047       53.0             1                 1      F   

  income_type  debt total_income_category                     purpose  
0   сотрудник     0                     B               покупка жилья  
1   сотрудник     0                     C     приобретение автомобиля  
2   сотрудник     0                     C               покупка жилья  
3   сотрудник     0                     B  дополнительное образование  
4   пенсионер     0                     C             сыграть свадьбу

### Категоризация целей кредита

Создаём функцию, которая выделит категории целей кредитования

In [50]:
def purpose_category (purpose): 
    """
    Возвращает категорию цели получения кредита по значению purpose,
    используя следующие правила:
    - 'операции с автомобилем', если в описании цели получения кредита указан автомобиль;
    - 'операции с недвижимостью', если в описании цели получения кредита указаны жильё и недвижимость;
    - 'проведение свадьбы', если в описании цели получения кредита - свадьба;
    - 'получение образования', если в описании цели получения кредита - образование.
    """
    if 'автомобил' in purpose:
        return 'операции с автомобилем'
    if 'недвижимост' in purpose:
        return 'операции с недвижимостью'
    if 'жиль' in purpose:
        return 'операции с недвижимостью'
    if 'свадьб' in purpose:
        return 'проведение свадьбы'
    if 'образовани' in purpose:
        return 'получение образования'
    

Проверяем работу функции на примерах:

In [51]:
purpose_category('на покупку автомобиля') 

'операции с автомобилем'

In [52]:
purpose_category('покупка жилья') 

'операции с недвижимостью'

Функция работает корректно.

Добавляем новый столбец с помощью применения новой функции:

In [53]:
data['purpose_category'] = data['purpose'].apply(purpose_category) 

Создаём словарь для данных о цели кредита и удаляем в нём дубликаты:

In [54]:
purpose_dict = data[['purpose_category', 'purpose']] 
purpose_dict = purpose_dict.drop_duplicates().reset_index(drop=True) 

In [55]:
data = data[['children','days_employed','dob_years','education_id','family_status_id','gender',
             'income_type','debt','total_income_category','purpose_category']]

Проверяем, как теперь выглядит таблица

In [56]:
display(data.head()) 

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1   -8437.673028       42.0             0                 0      F   
1         1   -4024.803754       36.0             1                 0      F   
2         0   -5623.422610       33.0             1                 0      M   
3         3   -4124.747207       32.0             1                 0      M   
4         0  340266.072047       53.0             1                 1      F   

  income_type  debt total_income_category          purpose_category  
0   сотрудник     0                     B  операции с недвижимостью  
1   сотрудник     0                     C    операции с автомобилем  
2   сотрудник     0                     C  операции с недвижимостью  
3   сотрудник     0                     B     получение образования  
4   пенсионер     0                     C        проведение свадьбы

### Вывод

In [57]:
len(data_raw) - len(data)

71

In [58]:
(len(data_raw) - len(data)) / len(data_raw) *100

0.32984901277584205

По результатам предобработки данных удалена 71 строка, это 0.33% от начального объёма данных.

Весь намеченный объём работ по предобработке данных выполнен. Датасет подготовлен для построения модели кредитного скоринга.

## Ответы на вопросы заказчика

### Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [59]:
data_children_pivot = data.pivot_table(index=['children'], values='debt', aggfunc=['count', 'sum', 'mean'])
data_children_pivot.columns = ['количество заёмщиков', 'количество должников', 'доля должников']

def percentage(i):
    return f"{i:.2%}"

data_children_pivot['доля должников'] = data_children_pivot['доля должников'].apply(percentage)

display(data_children_pivot.sort_values(by='доля должников', ascending=False))

количество заёмщиков  количество должников доля должников
children                                                           
4                           41                     4          9.76%
2                         2128                   202          9.49%
1                         4855                   445          9.17%
3                          330                    27          8.18%
0                        14091                  1063          7.54%
5                            9                     0          0.00%

### Вывод

Предоставленные данные говорят о том, что:
- доля должников без детей (7.54%) в целом чуть ниже, чем доля должников с детьми (от 8.19 до 9.76%) 
- 1 ребёнок у клиента или 2 детей - на возврат кредита в срок это не повлияет (доля должников у этих групп похожая - 9.17 и 9.49% соответственно)
- многодетных клиентов (с 3, 4, 5 детьми) на порядки меньше, и здесь более заметны колебания (доли должников 8.18%, 9.76%, 0% соответственно), хотя в среднем эти данные продолжают тенденцию более многочисленных групп клиентов с 1 и 2 детьми.

**Таким образом**, мы видим, что **клиенты с детьми чаще становятся должниками по сравнению с бездетными заёмщиками**. 

### Есть ли зависимость между семейным положением и возвратом кредита в срок? 

In [60]:
data_family = data.merge(family_status_dict, on='family_status_id', how='left')
data_family_pivot = data_family.pivot_table(index=['family_status'], values='debt', aggfunc=['count', 'sum', 'mean'])
data_family_pivot.columns = ['количество заёмщиков', 'количество должников', 'доля должников']
data_family_pivot['доля должников'] = data_family_pivot['доля должников'].apply(percentage)

display(data_family_pivot.sort_values(by='доля должников', ascending=False))

количество заёмщиков  количество должников  \
family_status                                                       
Не женат / не замужем                  2810                   274   
гражданский брак                       4151                   388   
женат / замужем                       12339                   931   
в разводе                              1195                    85   
вдовец / вдова                          959                    63   

                      доля должников  
family_status                         
Не женат / не замужем          9.75%  
гражданский брак               9.35%  
женат / замужем                7.55%  
в разводе                      7.11%  
вдовец / вдова                 6.57%

### Вывод


- Чаще всего должники встречаются среди клиентов, **не состоящих в браке** (статус "не женат/не замужем"), их доля составляет **9.75%**,
- Чуть меньше доля должников, состоящих **в гражданском браке**, это **9.35%**,
- Реже встречаются должники в категории клиентов, состоящих **в браке**, это **7.55%**,
- И реже всего встречаются должники среди клиентов **в разводе** и категорий **вдовец/вдова**, в этих группах по **7.11** и **6.57%** соответственно.

**Доля должников больше среди тех категорий пользователей, кто не связывал себя узами брака** и официально не регистрировал отношения, возможно, это в какой-то степени связано с личными качествами (ответственность).

**Возможно, здесь косвенно прослеживается взаимосвязь с возрастом клиентов** (ведь не состоят в браке или находятся в гражданском браке, как правило, люди молодые; а имеют за плечами развод или являются вдоцами/вдовами люди, имеющие более весомый жизненный опыт, и, соответственно, возраст). А возраст клиентов, в свою очередь, может быть связан с профессиональной квалификацией и более выгодным трудоустройством, более высоким доходом, а возможно и просто с более прагматичным, рациональным, взвешенным отношением к жизни. 

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [61]:
data_total_income_pivot = data.pivot_table(index=['total_income_category'], values='debt', aggfunc=['count', 'sum', 'mean'])
data_total_income_pivot.columns = ['количество заёмщиков', 'количество должников', 'доля должников']
data_total_income_pivot['доля должников'] = data_total_income_pivot['доля должников'].apply(percentage)

display(data_total_income_pivot.sort_values(by='доля должников', ascending=False))

количество заёмщиков  количество должников  \
total_income_category                                               
E                                        22                     2   
C                                     16015                  1360   
A                                        25                     2   
B                                      5042                   356   
D                                       350                    21   

                      доля должников  
total_income_category                 
E                              9.09%  
C                              8.49%  
A                              8.00%  
B                              7.06%  
D                              6.00%

### Вывод:

Данные показывают, что:

- наибольшая доля должников среди заёмщиков с категорией доходов **'E' (до 30 тыс.руб.)**, но здесь надо учитывать количество таких клиентов - их очень мало, это может влиять на результат
- высокий процент должников среди клиентов с категорией доходов **'C' (50-200 тыс.руб.)**, и это самая многочисленная категория
- клиенты с доходов **свыше 1 млн.руб. (категория доходов 'A')** тоже очень немногочисленны, но доля должников здесь ниже, чем у клиентов с категорией доходов 'E'(до 30 тыс.руб.) - это 8.00% против 9.09%
- доля должников среди заёмщиков с доходами категории **'B' (200 тыс.руб.-1 млн.руб.)** составляет 7.06%, и это вторая категория по численности
- наименьшая доля должников среди клиентов с категорией дохода **'D' (30-50 тыс.руб.)** - 6.00%, и эта группа не многочисленна


Таким образом, данные говорят, что **тенденция неравномерна**. С большой долей вероятности должником может стать как клиент, имеющий **самый низкий доход ('E' - 9.09%)**, но не сильно от него может отличаться и клиент с **самым высоким уровнем дохода ('A' - 8.00%)**. Также высоки шансы встретить должников в **самой многочисленной группе ('C'- 8.49%)**.

**Если же принять во внимание неравномерное распределение количества клиентов по группам** (ведь где-то мы видим всего пару десятков человек, а где-то несколько тысяч), и сравнить только две самые многочисленные категории ('C' и 'B'), то мы **увидим общую тенденцию: с увеличением дохода доля должников уменьшается (8.49% и 7.06%)**.

### Как разные цели кредита влияют на его возврат в срок?

In [62]:
data_purpose_pivot = data.pivot_table(index=['purpose_category'], values='debt', aggfunc=['count', 'sum', 'mean'])
data_purpose_pivot.columns = ['количество заёмщиков', 'количество должников', 'доля должников']
data_purpose_pivot['доля должников'] = data_purpose_pivot['доля должников'].apply(percentage)

display(data_purpose_pivot.sort_values(by='доля должников', ascending=False))

количество заёмщиков  количество должников  \
purpose_category                                                       
операции с автомобилем                    4306                   403   
получение образования                     4013                   370   
проведение свадьбы                        2324                   186   
операции с недвижимостью                 10811                   782   

                         доля должников  
purpose_category                         
операции с автомобилем            9.36%  
получение образования             9.22%  
проведение свадьбы                8.00%  
операции с недвижимостью          7.23%

### Вывод

Данные говорят:

- самая маленькая доля должников среди клиентов, чья цель - **операции с недвижимостью** - это **7.23%**
- на следующей ступени доля должников среди клиентов с целью **проведения свадьбы** - это **8.00%**
- и схожие доли должников среди клиентов, чьи цели связаны с  **получением образования** и **операциями с автомобилями** - **9.22%** и **9.36%** соответсвенно.

В этих категориях нет слишком кардинального разброса в количестве клиентов (везде счёт идёт на тысячи заёмщиков). Значит, эти показатели не являются "случайными".

**Могу предположить, что наиболее масштабные и долговременные цели (операции с недвижимостью) и цели, связанные с личной ответственностью (официальная регистрация брака и проведение свадьбы) побуждают заёмщиков быть более ответственными** и в своевременном погашении платежей по кредиту. 

## Общий вывод

По результатам предобработки данных удалена 71 строка, это 0.33% от начального объёма данных.
Весь намеченный объём работ по предобработке данных выполнен. Датасет подготовлен для построения модели кредитного скоринга.


Что касается вопросов заказчика: предоставленные данные говорят о том, что **количество детей, семейное положение, ежемесячный доход и цель кредита не слишком сильно влияют на погашение кредита в срок**. Данные по всем этим категориям **отличаются незначительно**, разброс идёт на несколько процентов (от 6.00 до 9.76%), **но некоторые закономерности**, тем не менее, **можно выявить**.


- **доля должников среди клиентов с детьми чуть выше** (от 8.19 до 9.76%), чем доля должников среди клиентов без детей (7.54%), а само количество детей на эту закономерность влияет незначительно и неравномерно: количество многодетных клиентов небольшое (и поэтому чёткую тенденцию на малой выборке выявить сложно);


- **чаще встречаются должники среди клиентов, не состоящих в браке и состоящих в гражданском браке**. Похоже, имеет место фактор личной ответственности (решение официально связать себя узами брака) или, возможно, косвенно влияют другие факторы (например, возраст, образование и доход);


- заёмщики неравномерно распределены по группам категорий доходов (в некоторых группах всего пара десятков клиентов, а в некоторых - это тысячи людей), поэтому **в целом взаимосвязь дохода и возвратом кредита в срок выглядит неравномерной** (сказывается малая выборка в нескольких категориях). Например, чаще всего должником становится клиент с категорией дохода 'E'(до 30 тыс.руб.) - 9.09%, но достаточно высокий процент должников и среди клиентов с самым высоким уровнем дохода - категория 'A'(свыше 1 млн.руб.) - 8.00%. При этом важно понимать, что всего клиентов с категорией доходов 'E' - 22, а с категорией 'A' - 25 человек. Если же посмотреть на самые многочисленные категории ('C' и 'B', где представлены тысячи клиентов), то мы **увидим общую тенденцию: с увеличением дохода доля должников уменьшается** (8.49% и 7.06%). Значит, **для банка менее благонадёжными выглядят заёмщики с небольшим уровнем дохода**.


- чаще не возвращают кредит в срок те клиенты, что оформляли его **для операций с автомобилями** и **получения образования**. Можно предположить, что наиболее масштабные и долговременные цели (операции с недвижимостью) и цели, связанные с личной ответственностью (официальная регистрация брака и проведение свадьбы) побуждают заёмщиков быть более ответственными.

Спасибо за внимание!
